tools/src/main/scala/scala/scalanative/linker/Reach.scala

```diff
@@ -138,6 +139,10 @@ class Reach(config: build.Config, entries: Seq[Global], loader: ClassLoader) {
             }
           }
         }
+        val clinit = cls.name.member(Sig.Clinit())
+        if (loaded(cls.name).contains(clinit)) {
+          reachGlobal(clinit)
+        }
```

Comment:
There's something fishy here. If you unroll the code above this snippet to see all of `def reachEntry`, you can see that when `reachEntry` is called, as soon as the class is not abstract, `reachAllocation(cls)` will be called, which will make instances of the class reachable. Calling `reachEntry` for classes with a static initializer will therefore cause too much methods to be reachable, because a non-abstract class with a static initializer will have its instances reachable, although the presence of a static initializer would not necessarily cause the class to be instantiated.

I believe you cannot reuse `reachEntry`. You need a different `def reachClinit` that focuses on reaching the static initializers, and not reach the allocation.